code/modules/asset_cache/transports/asset_transport.dm

```diff
@@ -0,0 +1,151 @@
+		OACI.namespace_parent = ACI.namespace_parent = (ACI.namespace_parent | OACI.namespace_parent)
+		OACI.namespace = OACI.namespace || ACI.namespace
+		if(OACI.hash != ACI.hash)
+			var/error_msg = "ERROR: new asset added to the asset cache with the same name as another asset: [asset_name] existing asset hash: [OACI.hash] new asset hash:[ACI.hash]"
+			stack_trace(error_msg)
+			log_debug(error_msg)
+		else
+			if(length(ACI.namespace))
+				return ACI
+			return OACI
+
+	SSassets.cache[asset_name] = ACI
+	return ACI
+
+
+/// Returns a url for a given asset.
+/// asset_name - Name of the asset.
+/// asset_cache_item - asset cache item datum for the asset, optional, overrides asset_name
+/datum/asset_transport/proc/get_asset_url(asset_name, datum/asset_cache_item/asset_cache_item)
+	if(!istype(asset_cache_item))
+		asset_cache_item = SSassets.cache[asset_name]
+	// To ensure code that breaks on cdns breaks in local testing, we only
+	// use the normal filename on legacy assets and name space assets.
+	var/keep_local_name = dont_mutate_filenames \
+		|| asset_cache_item.legacy \
+		|| asset_cache_item.keep_local_name \
+		|| (asset_cache_item.namespace && !asset_cache_item.namespace_parent)
+	if(keep_local_name)
+		return url_encode(asset_cache_item.name)
+	return url_encode("asset.[asset_cache_item.hash][asset_cache_item.ext]")
+
+
+/// Sends a list of browser assets to a client
+/// client - a client or mob
+/// asset_list - A list of asset filenames to be sent to the client. Can optionally be assoicated with the asset's asset_cache_item datum.
+/// Returns TRUE if any assets were sent.
+/datum/asset_transport/proc/send_assets(client/client, list/asset_list)
+	if(!istype(client))
+		if(ismob(client))
+			var/mob/M = client
+			if(M.client)
+				client = M.client
+			else //no stacktrace because this will mainly happen because the client went away
+				return
+		else
+			CRASH("Invalid argument: client: `[client]`")
+	if(!islist(asset_list))
+		asset_list = list(asset_list)
+	var/list/unreceived = list()
+
+	for(var/asset_name in asset_list)
+		var/datum/asset_cache_item/ACI = asset_list[asset_name]
+		if(!istype(ACI) && !(ACI = SSassets.cache[asset_name]))
+			log_debug("ERROR: can't send asset `[asset_name]`: unregistered or invalid state: `[ACI]`")
+			continue
+		var/asset_file = ACI.resource
+		if(!asset_file)
+			log_debug("ERROR: can't send asset `[asset_name]`: invalid registered resource: `[ACI.resource]`")
+			continue
+
+		var/asset_hash = ACI.hash
+		var/new_asset_name = asset_name
+		var/keep_local_name = dont_mutate_filenames \
+			|| ACI.legacy \
+			|| ACI.keep_local_name \
+			|| (ACI.namespace && !ACI.namespace_parent)
+		if(!keep_local_name)
+			new_asset_name = "asset.[ACI.hash][ACI.ext]"
+		if(client.sent_assets[new_asset_name] == asset_hash)
+			log_debug("Skipping send of `[asset_name]` (as `[new_asset_name]`) for `[client]` because it already exists in the client's sent_assets list")
```

Comment:
```suggestion
			// Un-comment below to debug asset sending (This will spam logs so do not enable normally)
			// log_debug("Skipping send of `[asset_name]` (as `[new_asset_name]`) for `[client]` because it already exists in the client's sent_assets list")
```